web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/IncomingEntitiesDataHelperTrait.php

```diff
@@ -0,0 +1,269 @@
+   *
+   * @param string $solution_id
+   *   The solution entity id.
+   *
+   * @return bool
+   *   Whether the solution id exists on the root level of the solution data.
+   */
+  protected function solutionDataRootExists(string $solution_id): bool {
+    $this->ensureEntityDataLoaded();
+    return isset($this->solutionData[$solution_id]);
+  }
+
+  /**
+   * Adds a solution id on the root of the solution data array.
+   *
+   * This method does not check if the root already exists and initializes the
+   * entry as a new array.
+   *
+   * @param string $solution_id
+   *   The solution entity id.
+   */
+  protected function addSolutionDataRoot(string $solution_id): void {
+    $this->ensureEntityDataLoaded();
+    $this->solutionData[$solution_id] = [];
+  }
+
+  /**
+   * Sets an entity as a dependency to the structured data.
+   *
+   * @param string $parent
+   *   The parent solution entity id.
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The child entity.
+   */
+  protected function addSolutionDataChildDependency(string $parent, EntityInterface $entity): void {
+    $this->ensureEntityDataLoaded();
+    $this->solutionData[$parent]['dependencies'][$entity->bundle()][$entity->id()] = $entity->id();
+  }
+
+  /**
+   * Returns whether the given solution has the given entity as a dependency.
+   *
+   * @param string $parent
+   *   The parent solution id.
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The candidate child entity.
+   *
+   * @return bool
+   *   Whether the solution already has this entity listed as a dependency.
+   */
+  protected function hasSolutionDataChildDependency(string $parent, EntityInterface $entity): bool {
+    $this->ensureEntityDataLoaded();
+    return isset($this->solutionData[$parent]['dependencies'][$entity->bundle()][$entity->id()]);
+  }
+
+  /**
+   * Returns a list of dependencies for a given solution.
+   *
+   * @param string $solution_id
+   *   The parent solution entity id.
+   *
+   * @return array
+   *   A structured array of entity ids listed as a dependency of the solution
+   *   indexed by their bundle.
+   */
+  protected function getSolutionDataChildDependencies(string $solution_id): array {
+    $this->ensureEntityDataLoaded();
+    return $this->solutionData[$solution_id]['dependencies'];
+  }
+
+  /**
+   * Returns a flat list of dependencies for a list of solutions.
+   *
+   * @param array $solution_ids
+   *   A list of solution ids.
+   *
+   * @return array
+   *   A flat list of dependencies concerning all solutions in the given list.
+   *   The flat list includes the solutions as well and adds the solutions and
+   *   the releases first in the list.
+   */
+  protected function getSolutionsWithDependenciesAsFlatList(array $solution_ids): array {
+    $this->ensureEntityDataLoaded();
+    $requested_dependencies = [];
+
+    $skipped_solutions = array_diff(array_keys($this->solutionData), $solution_ids);
```

Comment:
Actually, the main idea is that if a solution has another solution as a dependency, it might be that the second will end up in the list of dependencies without asking for it. However, as we are not adding a solution as a dependency to another solution, this is now redundant.
Removing.